frontend/src/scenes/trends/ActionFilter/actionFilterLogic.js

```diff
@@ -0,0 +1,138 @@
+import { kea } from 'kea'
+
+import { actionsModel } from '~/models/actionsModel'
+import { eventsModel } from '~/models/eventsModel'
+import { trendsLogic } from '../trendsLogic'
+
+import { groupActions, groupEvents } from '~/lib/utils'
+
+export const EntityTypes = {
+    ACTIONS: 'actions',
+    EVENTS: 'events',
+    NEW: 'new',
+}
+
+const mirrorValues = (entities, newKey, valueKey) => {
+    let newEntities = entities.map(entity => {
+        return {
+            ...entity,
+            [newKey]: entity[valueKey],
+        }
+    })
+    return newEntities
+}
+
+export const entityFilterLogic = kea({
+    connect: {
+        values: [eventsModel, ['events'], actionsModel, ['actions'], eventsModel, ['events'], trendsLogic, ['filters']],
+        actions: [trendsLogic, ['setFilters']],
+    },
+
+    actions: () => ({
+        selectFilter: filter => ({ filter }),
+        updateFilterMath: filter => ({ type: filter.type, value: filter.value, math: filter.math }),
+        updateFilter: filter => ({ type: filter.type, index: filter.index, value: filter.value }),
+        removeFilter: filter => ({ value: filter.value, type: filter.type, index: filter.index }),
+        createNewFilter: () => {},
+        removeNewFilter: () => {},
+    }),
+
+    reducers: ({ actions }) => ({
+        selectedFilter: [
+            null,
+            {
+                [actions.selectFilter]: (state, { filter }) => filter,
+            },
+        ],
+        newFilters: [
+            [],
+            {
+                [actions.createNewFilter]: state => [...state, { id: null }],
+                [actions.removeNewFilter]: state => {
+                    let newState = [...state]
+                    newState.pop()
+                    return newState
+                },
+            },
+        ],
+    }),
+
+    selectors: ({ selectors }) => ({
+        formattedFilters: [
+            () => [selectors.filters],
+            filters => {
+                return {
+                    [EntityTypes.ACTIONS]: filters.actions,
+                    [EntityTypes.EVENTS]: filters.events,
+                }
+            },
+        ],
+        entities: [
+            () => [selectors.events, selectors.actions],
+            (events, actions) => {
+                return {
+                    [EntityTypes.ACTIONS]: actions,
+                    [EntityTypes.EVENTS]: mirrorValues(events, 'id', 'name'),
+                }
+            },
+        ],
+        formattedOptions: [
+            () => [selectors.events, selectors.actions],
+            (events, actions) => {
+                return {
+                    [EntityTypes.ACTIONS]: groupActions(actions),
+                    [EntityTypes.EVENTS]: groupEvents(events),
+                }
+            },
+        ],
+    }),
+
+    listeners: ({ actions, values, props }) => ({
+        [actions.updateFilter]: ({ type, index, value }) => {
```

Comment:
This function still feels hard to reason about because it's so long (which means long blocks of code that aren't "named" in a way). There seems to be some overlap between the three functions so we could either find ways of reducing total lines, or split this up into three separate functions.

Reply:
yeah definitely. This bit along with the UX are slightly troubling because we're dealing with multiple types. Currently, the filterlogic depends on the trendlogic which provides it two distinct sets of filters (actions/events). Because of this, the ui displays them in clusters since we don't have an ordering. If we don't have the distinct groups on the ui, adding a new action filter would place a filter into the middle of the list (between the last action filter and first event filter). Going to work on a way to keep the ordering of the filters and it should be ok!